go/pools/smartconnpool/waitlist.go

```diff
@@ -50,61 +50,104 @@ type waitlist[C Connection] struct {
 // forced an expiration of all waiters in the waitlist.
-func (wl *waitlist[C]) waitForConn(ctx context.Context, setting *Setting) (*Pooled[C], error) {
+func (wl *waitlist[C]) waitForConn(ctx context.Context, setting *Setting, closeChan <-chan struct{}) (*Pooled[C], error) {
 	elem := wl.nodes.Get().(*list.Element[waiter[C]])
+	defer wl.nodes.Put(elem)
 	elem.Value = waiter[C]{setting: setting, conn: nil, ctx: ctx}
 
 	wl.mu.Lock()
 	// add ourselves as a waiter at the end of the waitlist
 	wl.list.PushBackValue(elem)
 	wl.mu.Unlock()
 
-	// block on our waiter's semaphore until somebody can hand over a connection to us
-	elem.Value.sema.wait()
+	done := make(chan struct{})
+	go func() {
+		// Block on our waiter's semaphore until somebody can hand over a connection to us.
+		elem.Value.sema.wait()
+		close(done)
+	}()
+
+	select {
+	case <-closeChan:
+		// Pool was closed while we were waiting.
+		removed := false
+
+		wl.mu.Lock()
+		// Try to find and remove ourselves from the list.
+		for e := wl.list.Front(); e != nil; e = e.Next() {
+			if e == elem {
+				wl.list.Remove(elem)
+				removed = true
+				break
+			}
+		}
+		wl.mu.Unlock()
+
+		// If we removed ourselves from the waitlist, we need to notify our semaphore
+		if removed {
+			elem.Value.sema.notify(false)
+		}
+
+		// Wait for the semaphore to have been notified, either by us or by someone else
+		<-done
+
+		if removed {
+			return nil, ErrConnPoolClosed
+		}
+
+		return elem.Value.conn, nil
+
+	case <-ctx.Done():
+		// Context expired. We need to try to remove ourselves from the waitlist to
+		// prevent another goroutine from trying to hand us a connection later on.
+		removed := false
+
+		wl.mu.Lock()
+		// Try to find and remove ourselves from the list.
+		for e := wl.list.Front(); e != nil; e = e.Next() {
+			if e == elem {
+				wl.list.Remove(elem)
+				removed = true
+				break
+			}
+		}
+		wl.mu.Unlock()
+
+		// If we removed ourselves from the waitlist, we need to notify our semaphore
+		if removed {
+			elem.Value.sema.notify(false)
+		}
 
-	// we're awake -- the conn in our waiter contains the connection that was handed
-	// over to us, or nothing if we've been waken up forcefully. save the conn before
-	// we return our waiter to the pool of waiters for reuse.
-	conn := elem.Value.conn
-	wl.nodes.Put(elem)
+		// Wait for the semaphore to have been notified, either by us or by someone else
+		<-done
 
-	if conn != nil {
-		return conn, nil
+		if removed {
+			return nil, context.Cause(ctx)
+		}
+
+		return elem.Value.conn, nil
+
+	case <-done:
+		return elem.Value.conn, nil
 	}
-	return nil, ctx.Err()
 }
 
-// expire removes and wakes any expired waiter in the waitlist.
-// if force is true, it'll wake and remove all the waiters.
-func (wl *waitlist[C]) expire(force bool) (maybeStarving int) {
+func (wl *waitlist[C]) maybeStarvingCount() (maybeStarving int) {
 	if wl.list.Len() == 0 {
 		return
 	}
 
-	var expired []*list.Element[waiter[C]]
-
```

Comment:
I'm not sure why we're removing the expiration code in the PR? Is it no longer needed?